content/blogposts/2024/anatomy-of-a-coding-assistant.md

```diff
@@ -0,0 +1,166 @@
+
+![6_edits](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/6_edits.png)
+
+Each of these features works very differently from one another. The mental model of the user and the requirements are different for each. For autocomplete, speed is key. You want it to be very fast and accurate enough so that it doesn’t interrupt the programmer’s flow. On the other hand, for chat, accuracy and the quality of the responses may be more important.
+
+![7_tradeoff](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/7_tradeoff.png)
+
+When generating tests, there may be additional requirements. For example, if you’re looking at a file and want to add tests to it, you need to quickly detect if a test file for that file already exists. So, when we say “coding assistant,” it might sound simple, but it’s actually a product of products—multiple products put together to create a comprehensive experience.
+
+In this post, we're going to take a look at how context fetching works for each particular feature - chat, autocomplete, generating tests, and editing code.
+
+---
+
+## How context fetching works for chat
+
+![8_chat](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/8_chat.png)
+
+Chat is the feature that allows you to interact with an AI model about your codebase. For this, we want a few things:
+
+1. **Conversation history:** We need to keep track of the entire conversation history so it makes sense as continuous chat.
+2. **Automatic context fetching:** It's important to automatically fetch the most relevant code snippets related to the user's query.
+3. **User control:** Users should have the ability to mention specific files and provide those directly to the model, and also include the option to reference external sources like Slack threads or Notion documents to enrich the context further.
+
+![9_context](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/9_context.png)
+
+For automatic context fetching, we’ve found that there are two options that work fairly well:
+
+1. **Indexed search:** We ask an LLM a question like, "Given this user query, what keywords should be used to search for relevant parts of the codebase?" Once the AI provides keywords, we use a search engine that leverages indexing to perform the search. We get search rankings, and based on a predefined token limit, we send the most relevant search results back to the main AI model you're having a conversation with.
+    1. For free and Pro users, we use our home-grown search engine called symf, written in Golang, that runs on your local environment.
+    2. For Enterprise users, we leverage [Sourcegraph’s Code Search](https://sourcegraph.com/code-search) for this.
```

Comment:
```suggestion
```

Reply:
this is going to change soon and at a high level, both work the same, so we should omit these 2 lines